Optional web-specific deps for wasm32

raphaelhetzel opened the pull request:
**Connections**
Related to what lead to #7533, but idependent.

**Description**
In the current version, the target_plaform wasm32 (without target_os emscripten) depends on web-specific dependencies. However, wasm can be used outside of the web, and wgpu can be used on wasm runtimes without javascript bindings / the webgpu backend using a custom backend.

**Testing**
I haven't tested emscripten or webgl (There is an idependent error when building webgl).
There might be some issues people working with emscripten/webgl might be able to detect here.
I am also not sure whether this could be simplified (the not emscripten in wgpu/Cargo.toml might be unnecessary).

**Squash or Rebase?**

Squash

**Checklist**

- [x ] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [ x] Run `cargo clippy --tests`. If applicable, add:
  - [ x] `--target wasm32-unknown-unknown` (some independent warnings)
- [X ] Run `cargo xtask test` to run tests. (one naga error I don't think is related to this)
- [ ] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

raphaelhetzel committed:
Optional web-specific deps for wasm32

cwfitzgerald reviewed (changes requested):
Looks good! Before we land, as dependencies are really tricky and easy to regress, could you add [dependency tests](https://github.com/gfx-rs/wgpu/blob/trunk/tests/tests/wgpu-dependency/main.rs) for:
- Ensuring that depending on wgpu on wasm with just the custom feature does not depend on wasm-bindgen et al. 
- Ensuring that turning on either the webgl or webgpu feature does add wasm-bindgen et al.
- Ensuring turning on both ^ does not bring in wasm-bindgen on a native platform (lets say windows)

raphaelhetzel committed:
Dependency tests for web-specific bindings

cwfitzgerald commented on tests/tests/wgpu-dependency/main.rs:
```diff
@@ -121,7 +121,7 @@ fn wasm32_with_webgpu_and_wgsl_does_not_depend_on_naga() {
         target: "wasm32-unknown-unknown",
         packages: &["wgpu"],
-        features: &["webgpu", "wgsl"],
+        features: &["custom"],
```
This change seems errant?

raphaelhetzel replied:
Yes, sorry for that, I must have slipped up while adding the tests below and did not notice it during the push.

cwfitzgerald reviewed (changes requested):
One issue, one question. Should we still have a `web` feature on `wgpu` as well to enable it with a custom backend running on the web? Might not be worth the bother.

raphaelhetzel commented:
Regarding your question, I don't see what the `web` feature would enable.
There are features for all backends available on wasm: Custom does not require web-sys et. al., while webgpu and webgl require those dependencies. I have not added that test case, but there is no reason why all three can't be enabled at the same time.

We could add a web feature to semantically group the dependency enables instead of duplicating them for the two backends.

cwfitzgerald commented:
The case I was thinking of was: you want a custom backend that runs on web-enabled wasm and want the web-sys types in `wgpu-types`. I guess you can just enable `wgpu-types/web` at that point?

sagudev commented on wgpu/Cargo.toml:
```diff
@@ -68,7 +72,14 @@ angle = ["wgpu-core?/angle"]
 ## Enables the GLES backend on WebAssembly only.
-webgl = ["wgpu-core/webgl", "dep:wgpu-hal", "dep:smallvec"]
+webgl = [
+    "wgpu-core/webgl",
```
`wgpu-core/webgl` does enables `wgpu-types/web`, but I am wondering if would be worth being explicit about this:
```suggestion
    "wgpu-core/webgl",
    "wgpu-types/web",
```

raphaelhetzel committed:
Add 'web' feature, add missing feature enables/checks.

raphaelhetzel commented:
I've added a `web` feature now. That also covers @sagudev's comment.
I also added some additional enables and a check for the custom_backend example.

raphaelhetzel commented:
I am not quite sure how I should go about fixing the tests failing due to the custom example.
The standalone examples depend on wgpu with the default parameters and therefore enable the `web` feature.
The CI runs however disable the default features and therefore disable the web feature for the custom example.

This would work if only the custom example ist built.

I guess I would either need to always depend on `web` for the custom_example or change the pipeline.
It should not have worked prior to todays commit but did by accident.

sagudev commented:
Maybe just tweak CI invocations by adding `web` feature on wasm?

sagudev commented:
You can take some inspiration from: https://github.com/gfx-rs/wgpu/pull/7407/files

raphaelhetzel commented:
Thanks for the hint. Wasn't quite sure whether that would be acceptable (essentially I could also just change that for that one example). But as I am not sure whether anyone else will actually use this, I did not want to add a separate matrix entry for wasm with and without web specifics.

sagudev commented:
> Thanks for the hint. Wasn't quite sure whether that would be acceptable (essentially I could also just change that for that one example). But as I am not sure whether anyone else will actually use this, I did not want to add a separate matrix entry for wasm with and without web specifics.

The root problem is https://github.com/gfx-rs/wgpu/issues/7314

cwfitzgerald reviewed (approved):
This is wonderful stuff!

cwfitzgerald commented:
I'm going to update the branch to make sure CI is still clean, then we can land.

cwfitzgerald committed:
Merge branch 'trunk' into web_dependencies
